core/src/main/java/com/linecorp/armeria/common/stream/StreamWriterOutputStream.java

```diff
@@ -0,0 +1,114 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpData;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+final class StreamWriterOutputStream<T> extends OutputStream {
+
+    private final StreamWriter<T> writer;
+    private final Function<? super HttpData, ? extends T> httpDataConverter;
+    private final ByteBuf buffer = Unpooled.buffer();
+    private final int maxBufferSize;
+    private boolean closed;
+
+    StreamWriterOutputStream(StreamWriter<T> writer,
+                             Function<? super HttpData, ? extends T> httpDataConverter,
+                             int maxBufferSize) {
+        requireNonNull(writer, "writer");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        checkArgument(maxBufferSize > 0, "maxBufferSize should be positive");
+        this.writer = writer;
+        this.httpDataConverter = httpDataConverter;
+        this.maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public void write(int b) throws IOException {
+        ensureOpen();
+        maybeDrain();
+        buffer.writeByte(b);
+    }
+
+    @Override
+    public void write(byte[] b, int off, int len) throws IOException {
+        ensureOpen();
+        while (len > 0) {
+            maybeDrain();
+            final int writableBytes = maxBufferSize - buffer.readableBytes();
+            if (writableBytes < len) {
+                buffer.writeBytes(b, off, writableBytes);
+                off += writableBytes;
+                len -= writableBytes;
+            } else {
+                buffer.writeBytes(b, off, len);
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void flush() throws IOException {
+        ensureOpen();
+        drain();
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        flush();
+        writer.close();
+        closed = true;
+    }
+
+    private void ensureOpen() throws IOException {
+        if (closed) {
+            throw new IOException("Stream closed");
+        }
+        if (!writer.isOpen()) {
+            closed = true;
+            throw new IOException("Stream closed");
+        }
+    }
+
+    private void maybeDrain() {
+        if (buffer.readableBytes() >= maxBufferSize) {
+            drain();
+        }
+    }
+
+    private void drain() {
+        if (!buffer.isReadable() || !writer.isOpen()) {
+            return;
+        }
+        final byte[] data = new byte[buffer.readableBytes()];
+        buffer.readBytes(data);
+        writer.write(httpDataConverter.apply(HttpData.wrap(data)));
```

Comment:
- If we don't reset the `readerIndex()` and `writerIndex()`, the size of the backing array will continue to grow which can cause an `OutOfMemory` error.
- Should we wrap the `ByteBuf` with `HttpData` and pass it to `StreamWriter` and allocate a new `ByteBuf`?

Reply:
aha.. I understood your concern here. I didn't know that `OutOfMemory` can be occurred here.
How about just `buffer.clear()` on `drain()` ? (allocating new `ByteBuf` looks also good to me.)

Reply:
It would be efficient to wrap the filled `ByteBuf` with `HttpData` and pass it to `httpDataConverter` to avoid additional array copies.
After that, how about allocating a new `ByteBuf`.

Reply:
> to avoid additional array copies.

Aha!! I understood your intention here.
✅ updated, wrap `ByteBuf` and pass it & allocation a new `ByteBuf` on drain() to avoid additional array copies.